File: Algorithms/3. Graph Traversal/Breadth First Search/cpp/breadth_first_search.cpp
#include "breadth_first_search.hpp"

inline std::vector<TreeNode*> bfs(TreeNode* root) {
  TreeNode* node;
  std::vector<TreeNode*> children;
  std::vector<TreeNode*> queue { root };
  std::vector<TreeNode*> order { root };

  while (!queue.empty()) {
    node = queue[0];
    queue.erase(queue.begin());
    children = node->children();

    for (int index = 0; index < children.size(); index++) {
      queue.push_back(children[index]);
      order.push_back(children[index]);
    }
  }

  return order;
}

File: Algorithms/3. Graph Traversal/Breadth First Search/python/breadth_first_search.py
import sys
from pathlib import Path
sys.path.insert(0, str(Path(__file__).resolve().parents[4]/ "Data Structures" / "3. Trees" / "Binary Search Tree" / "python"))
from tree_node import TreeNode
from collections import deque

# log(n)
def breadth_first_search(root: TreeNode, value: int) -> TreeNode:
    queue = deque([root])

    while queue:
        node = queue.popleft()
        if node.value == value: return node
        if node.left != None: queue.append(node.left)
        if node.rite != None: queue.append(node.rite)

    return None
